### PR TITLE
Features/my journey changes

### PR DESCRIPTION
**Pull request by praveenKDass**
# Description
These recommendations are intended to promote code quality and team communication during software development. They cover a variety of topics, including ensuring that pull requests are submitted to the correct branch, documenting new methods, preserving consistency across many services, and avoiding typical blunders like accessing APIs or DB queries within loops. Sensitive data should not be uploaded, and changes to environment variables or database models should be executed consistently. Teams may work more effectively and develop higher-quality software by adhering to these standards.

## Type of change
Please choose appropriate options.

- [ ]  Bug fix (non-breaking change which fixes an issue)
- [ ]  New feature (non-breaking change which adds functionality)
- [ ]  Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ]  Enhancement (additive changes to improve performance)
- [ ]  This change requires a documentation update

## Checklist 

- [ ]  It's critical to avoid making needless file modifications in contributions, such as adding new lines, console logs, or additional spaces, to guarantee cleaner and more efficient code. Furthermore, eliminating unnecessary imports from a file might enhance code readability and efficiency.
- [ ]  Ensure that the pull request is assigned to the right base branch and that the development branch name contains the JIRA Task Id. Furthermore, each commit message should include the JIRA Task Id in the manner "ED-100: message".
- [ ]  Only update packages if it is mentioned and authorized in the design document, and make sure that you have the required permissions.
- [ ]  Avoid making API and database queries inside a loop as it can lead to performance issues and slow down the system.
- [ ]  When calling another function inside a given function, add comments explaining the purpose and meaning of the passed arguments and expected return values.
- [ ]  If adding a blank argument in a function, add a comment explaining the reason for the blank argument.
- [ ]  Before submitting a pull request, do a self-review of your code to ensure there are no conflicts with the base branch and all comments have been addressed.
- [ ]  Before merging a pull request, it's important to have other team members review it to catch any potential errors or issues
- [ ]  To maintain code integrity, it's important to remove all related changes when removing code during a code review.
- [ ]  If new constants, endpoints, or utility functions are introduced, it is important to check if they already exist in the service to avoid any duplication.
- [ ]  Whenever a new environment variable is added to a service, it's important to ensure that the necessary changes are made to related files such as ".env.sample" and "envVariables.js" to maintain consistency and avoid errors. Additionally, the new environment variable should be added to the devops repository to ensure that it is properly documented and accessible to the team.
- [ ]  When adding a new function to a service, it is important to document it with relevant information such as the name, parameters, and return value in a consistent format across all services. Additionally, if there are any changes to the API response, ensure that the documentation in the controllers is updated accordingly. 
- [ ]  Write a clear and concise commit message that describes the changes made.
- [ ]  Maintain consistent function signature and code across all services when adding a function to multiple services. Implement changes to database models in all services that use the same model. 
- [ ]  Use only let and const. Do not use var.
- [ ]  Make common functions for repetitive code blocks.
- [ ]  Avoid uploading sensitive information such as secret tokens or passwords in pull requests to ensure data security.
- [ ]  Maintain consistent indentation and spacing throughout the code.

### CODE REVIEW

File: controllers/v1/users.js [hunk 1]
```diff
@@ -149,11 +149,13 @@ module.exports = class Users {
 		return new Promise(async (resolve, reject) => {
 			try {
 				let isAPrivateProgram = UTILS.convertStringToBoolean(req.query.isAPrivateProgram)
+				req.query.isProjectsCount = UTILS.convertStringToBoolean(req.query.isProjectsCount)
```
Comment (VISHNUDAS-tunerlabs) [inline controllers/v1/users.js:152]: @praveenKDass we can call the query getProjectsCount

Comment (VISHNUDAS-tunerlabs) [inline controllers/v1/users.js:152]: @praveenKDass update the function signature

File: module/programs/helper.js [hunk 2]
```diff
@@ -1045,10 +1046,11 @@ module.exports = class ProgramsHelper {
...
 	 * @param {String} language -languageCode
+	 * @param {Boolean} projectsCount - get the projectsCount under that program.
 	 * @returns {JSON} - List of programs that user created on app.
 	 */
 
-	static userPrivatePrograms(userId, language = '') {
+	static userPrivatePrograms(userId, language = '', projectsCount = false) {
 		return new Promise(async (resolve, reject) => {
 			try {
 				let programsData = await programsQueries.programsDocument(
```
Comment (VISHNUDAS-tunerlabs) [inline module/programs/helper.js:1056]: @praveenKDass check if we can apply isDeleted flag as well

File: module/programs/helper.js [hunk 3]
```diff
@@ -1067,6 +1069,46 @@ module.exports = class ProgramsHelper {
 						result: [],
 					})
 				}
+				if (projectsCount) {
+					//Filtering out all the program IDs
+					let allProjectIds = programsData.flatMap((program) => program._id || [])
```
Comment (VISHNUDAS-tunerlabs) [inline module/programs/helper.js:1074]: @praveenKDass rename the varable to userProgramIds

File: module/programs/helper.js [hunk 4]
```diff
@@ -1067,6 +1069,46 @@ module.exports = class ProgramsHelper {
...
+					let projectDocuments = await projectQueries.getAggregate([
+						{
+							$match: {
+								programId: { $in: allProjectIds },
```
Comment (VISHNUDAS-tunerlabs) [inline module/programs/helper.js:1078]: @praveenKDass change the variable name here as well

File: module/programs/helper.js [hunk 5]
```diff
@@ -1067,6 +1069,46 @@ module.exports = class ProgramsHelper {
...
+						acc[project._id.toString()] = {
+							ongoingCount: project.ongoingCount || 0,
+							completedCount: project.completedCount || 0,
+						}
+						return acc
+					}, {})
+
+					// Merge counts using the map
+					programsData.forEach((program) => {
+						const counts = projectCountsMap[program._id.toString()] || {
```
Comment (VISHNUDAS-tunerlabs) [inline module/programs/helper.js:1104]: @praveenKDass lets call them ongoingProjects and completedProjects

**Review by VISHNUDAS-tunerlabs — CHANGES_REQUESTED** [review]
Reviewed on Dec 4

File: controllers/v1/users.js (hunk 1 above)
Comment (praveenKDass) [reply controllers/v1/users.js:152]: Changed @VISHNUDAS-tunerlabs

Comment (praveenKDass) [reply controllers/v1/users.js:152]: Added @VISHNUDAS-tunerlabs

File: module/programs/helper.js (hunk 2 above)
Comment (praveenKDass) [reply module/programs/helper.js:1056]: added query based on isDeleted as well @VISHNUDAS-tunerlabs

File: module/programs/helper.js (hunk 3 above)
Comment (praveenKDass) [reply module/programs/helper.js:1074]: Changed @VISHNUDAS-tunerlabs

File: module/programs/helper.js (hunk 5 above)
Comment (praveenKDass) [reply module/programs/helper.js:1104]: Changed @VISHNUDAS-tunerlabs

File: controllers/v1/users.js [hunk 6]
```diff
@@ -142,18 +142,21 @@ module.exports = class Users {
...
 		return new Promise(async (resolve, reject) => {
 			try {
 				let isAPrivateProgram = UTILS.convertStringToBoolean(req.query.isAPrivateProgram)
+				req.query.getProjectsCount = UTILS.convertStringToBoolean(req.query.getProjectsCount)
 
 				if (isAPrivateProgram) {
 					let programsData = await usersHelper.privatePrograms(
 						req.userDetails.userInformation.userId,
-						req.query.language ? req.query.language : ''
+						req.query.language ? req.query.language : '',
```
Comment (VISHNUDAS-tunerlabs) [inline controllers/v1/users.js:158]: @praveenKDass Please introduce pagination to this API

**Review by VISHNUDAS-tunerlabs — CHANGES_REQUESTED** [review]

File: controllers/v1/users.js (hunk 6 above)
Comment (praveenKDass) [reply controllers/v1/users.js:158]: @VISHNUDAS-tunerlabs  added

File: generics/helpers/utils.js [hunk 7]
```diff
@@ -732,6 +732,39 @@ function getTranslatedData(data, translateData) {
...
+
+/**
+ * Function to calculate the end date based on a start date and duration string.
+ * @param {string} createdDate - The start date in ISO format (e.g., '2024-12-05T00:00:00.000Z').
+ * @param {string} durationString - Duration in the format "1 week", "2 weeks", "1 month", "1 year", etc.
+ * @returns {string} - The calculated end date in ISO format
+ */
+function calculateEndDate(createdDate, durationString) {
+	const startDate = moment(createdDate) // Parse the start date
+	console.log(`Calculating end date`, startDate)
```
Comment (VISHNUDAS-tunerlabs) [inline generics/helpers/utils.js:744]: @praveenKDass please remove this console

File: module/programs/helper.js [hunk 8]
```diff
@@ -1075,6 +1136,8 @@ module.exports = class ProgramsHelper {
 						externalId: program.externalId,
 						_id: program._id,
 						isAPrivateProgram: program.isAPrivateProgram,
+						ongoingProjects: program.ongoingProjects ? program.ongoingProjects : 0,
+						completedCount: program.completedCount ? program.completedCount : 0,
```
Comment (VISHNUDAS-tunerlabs) [inline module/programs/helper.js:1140]: @praveenKDass please make this completedProjects

**Review by VISHNUDAS-tunerlabs — CHANGES_REQUESTED** [review]

File: module/programs/helper.js (hunk 8 above)
Comment (praveenKDass) [reply module/programs/helper.js]: Changed @VISHNUDAS-tunerlabs

File: generics/helpers/utils.js (hunk 7 above)
Comment (praveenKDass) [reply generics/helpers/utils.js]: Removed @VISHNUDAS-tunerlabs

**Review by VISHNUDAS-tunerlabs — APPROVED** [review]

**Pull request merged**